esp-hal/src/dma/mod.rs

```diff
@@ -2338,13 +2419,10 @@ impl DmaRxBuf {
             let amount_to_copy = min(desc.len(), remaining.len());
 
             let (to_fill, to_remain) = remaining.split_at_mut(amount_to_copy);
-            to_fill.copy_from_slice(&self.buffer[buffer_offset..][..amount_to_copy]);
+            to_fill.copy_from_slice(&internal[..amount_to_copy]);
             remaining = to_remain;
 
-            if desc.next.is_null() {
-                break;
-            }
-            buffer_offset += desc.flags.size() as usize;
+            internal = &internal[amount_to_copy..];
```

Comment:
Incorrect, you want the `size()` not the `length()` (`amount_to_copy`).

Reply:
I wonder if this matters outside of burst mode, but also just recreating a subslice of the buffer from the descriptor is less error prone.

Reply:
> I wonder if this matters outside of burst mode

It does. When the DMA gets an EOF from the peripheral it closes the descriptor immediately and jumps to the next one.
The DMA may also close the descriptor early if the restart bit in the inlink register is set.

> just recreating a subslice of the buffer from the descriptor is probably less error prone.

I've been contemplating this myself but in general for the bufs. I think we might be breaking aliasing rules by keeping an exclusive reference to the buffer when the DMA is actually modifying it behind the scenes, leading to the compiler mis-optimising the code maybe.

Reply:
> I've been contemplating this myself but in general for the bufs. I think we might be breaking aliasing rules by keeping an exclusive reference to the buffer when the DMA is actually modifying it behind the scenes, leading to the compiler mis-optimising the code maybe.

Yeah we probably want to store pointers instead of the slice, but in practice we don't access the same data at the same time through two mutable references so this should be somewhat safe.